From bd5f02c51de262e76fbf547d21cf1e95a6e0c20e Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Thu, 23 Jun 2016 17:02:25 +0300
Subject: [PATCH 0292/1345] pp3: change MAC addr while interface down

commit  fdfd43fdf2e1cdae2309dbf3581cdd5857af8bf7 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Fix for SYSTEMSW-2656
Changing mac address while interface down
cause to segmentation fault

Change-Id: I5e528de0107ee3338d4cbb1dbb245051d6554dab
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30700
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Tested-by: Dmitri Epshtein <dima@marvell.com>
Verified-Armada38x-39x: Dmitri Epshtein <dima@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../net/ethernet/marvell/pp3/net_dev/mv_netdev.c   |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
index 20d8917..2920f65 100644
--- a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
+++ b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
@@ -1897,6 +1897,13 @@ int mv_pp3_set_mac_addr(struct net_device *dev, void *p)
 	if (!is_valid_ether_addr(addr->sa_data))
 		return -EADDRNOTAVAIL;
 
+	if (!dev_priv->vport) {
+		if (!dev->dev_addr)
+			goto error;
+		memcpy(dev->dev_addr, addr->sa_data, MV_MAC_ADDR_SIZE);
+		return 0;
+	}
+
 	if (dev_priv->vport->type != MV_PP3_NSS_PORT_ETH) {
 		memcpy(dev->dev_addr, addr->sa_data, MV_MAC_ADDR_SIZE);
 		return 0;
-- 
1.7.9.5

